upup/pkg/fi/assettasks/copyfile.go

```diff
@@ -0,0 +1,203 @@
+	sourceSHALocation := fi.StringValue(e.SourceShaLocation)
+
+	glog.Infof("copying bits from %q to %q", source, target)
+
+	if err := transferFile(source, target, sourceSha, sourceSHALocation); err != nil {
+		return fmt.Errorf("unable to transfer %q to %q: %v", source, target, err)
+	}
+
+	return nil
+}
+
+func transferFile(source string, target string, sourceSHA string, sourceSHALocation string) error {
+	data, err := vfs.Context.ReadFile(source)
+	if err != nil {
+		return fmt.Errorf("Error unable to read path %q: %v", source, err)
+	}
+
+	uploadVFS, err := buildVFSPath(target)
+	if err != nil {
+		return err
+	}
+
+	// Test the file matches the sourceSha
+	{
+		var sha string
+		if sourceSHA != "" {
+			sha = sourceSHA
+		} else if sourceSHALocation != "" {
+			shaBytes, err := vfs.Context.ReadFile(sourceSHALocation)
+			if err != nil {
+				return fmt.Errorf("Error unable to read path %q: %v", source, err)
+			}
+			sha = string(shaBytes)
+		}
+
+		if sha != "" {
+
+			trimmedSHA := strings.TrimSuffix(sha, "\n")
+
+			in := bytes.NewReader(data)
+			dataHash, err := hashing.HashesForResource(in, []hashing.HashAlgorithm{hashing.HashAlgorithmSHA1})
+			if err != nil {
+				return fmt.Errorf("unable to hash sha from data: %v", err)
+			}
+
+			shaHash, err := hashing.FromString(trimmedSHA)
+			if err != nil {
+				return fmt.Errorf("unable to hash sha: %q, %v", sha, err)
+			}
+
+			if !shaHash.Equal(dataHash[0]) {
+				return fmt.Errorf("SHAs are not matching for %q", dataHash[0].String())
+			}
+
+			shaVFS, err := buildVFSPath(target + ".sha1")
+			if err != nil {
+				return err
+			}
+
+			b := bytes.NewBufferString(sha)
+			if err := writeFile(shaVFS, b.Bytes()); err != nil {
+				return fmt.Errorf("Error uploading file %q: %v", shaVFS, err)
+			}
+		}
+	}
+
+	if err := writeFile(uploadVFS, data); err != nil {
+		return fmt.Errorf("Error uploading file %q: %v", uploadVFS, err)
+	}
+
+	return nil
+}
+
+func writeFile(vfsPath string, data []byte) error {
+	glog.V(2).Infof("uploading to %q", vfsPath)
+	destinationRegistry, err := vfs.Context.BuildVfsPath(vfsPath)
+	if err != nil {
+		return fmt.Errorf("Error parsing registry path %q: %v", vfsPath, err)
+	}
+
+	if err = destinationRegistry.WriteFile(data); err != nil {
+		return fmt.Errorf("Error destination path %q: %v", vfsPath, err)
+	}
+
+	glog.V(2).Infof("upload complete: %q", vfsPath)
+
+	return nil
+}
+
+func buildVFSPath(target string) (string, error) {
+	if !strings.Contains(target, "://") || strings.HasPrefix(target, "memfs://") {
+		return target, nil
+	}
+
+	u, err := url.Parse(target)
+	if err != nil {
+		return "", fmt.Errorf("unable to parse: %q", target)
+	}
+
+	var vfsPath string
+
+	// remove the filename from the end of the path
+	//pathSlice := strings.Split(u.Path, "/")
+	//pathSlice = pathSlice[:len(pathSlice)-1]
+	//path := strings.Join(pathSlice, "/")
+
+	// TODO I am not a huge fan of this, but it would work
+	// TODO @justinsb should we require an URL?
```

Comment:
Yes, I would just have the task expect "real" vfs path strings, i.e. ones that would work with `vfs.Context.BuildVfsPath`.  From a `vfs.Path` you can get the path using `Path()`